synfig-studio/src/gui/dialogs/dialog_pluginmanager.cpp

```diff
@@ -188,90 +247,51 @@ Dialog_PluginManager::on_install_plugin_button_clicked()
-                    zip_fs->directory_scan(file, child_files);
-                    for (auto& innerFile : child_files) {
-                        if (innerFile == "plugin.xml") {
-                            pluginStatus = PluginZipStatus::PLUGIN_ZIP_AT_CHILD_DIR;
-                            break;
-                        }
-                    }
-                }
-            }
-            std::string output_path ;
-            if (pluginStatus == PluginZipStatus::PLUGIN_ZIP_AT_ROOT_DIR) {
-                output_path = ((path_to_user_plugins / filesystem::Path(zip_filename).stem()).add_suffix("/")).u8string();
-                if (native_fs->directory_create(output_path))
-                for (const auto &file : files)
-                    zip_fs->copy_recursive(zip_fs, file, native_fs, output_path + file);
-                App::plugin_manager.load_plugin(output_path + "plugin.xml", output_path, true);
-            }
-            else if (pluginStatus == PluginZipStatus::PLUGIN_ZIP_AT_CHILD_DIR) {
-                std::string dir_name;
-                std::vector<std::string> child_files;
-                for (auto& file : files) {
-                    if (zip_fs->is_directory(file)) {
-                        zip_fs->directory_scan(file, child_files);
+                    if (zip_fs->directory_scan(file, child_files)) {
                         for (auto& innerFile : child_files) {
                             if (innerFile == "plugin.xml") {
-
-                                xmlpp::DomParser parser;
-                                parser.set_substitute_entities();
-		                        parser.parse_stream(*zip_fs->get_read_stream(file + '/' + innerFile));
-
-                                const xmlpp::Node* pNode = parser.get_document()->get_root_node();
-                                if (std::string(pNode->get_name()) != std::string("plugin") ) {
-                                    synfig::warning("Invalid plugin.xml file (missing root <plugin>)");
-                                    return;
-                                }
-
-                                auto execlist = pNode->find("./exec");
-                                if (!execlist.empty())
-                                {
-                                    for (const xmlpp::Node* node : pNode->find("./name") ) {
-                                        const xmlpp::Element* element = dynamic_cast<const xmlpp::Element*>(node);
-                                        std::string lang = element->get_attribute_value("lang");
-                                        if (lang.empty()) {
-                                            dir_name = element->get_child_text()->get_content();
-                                            break;
-                                        }
-                                    }
-                                }
+                                pluginStatus = PluginZipStatus::PLUGIN_ZIP_AT_CHILD_DIR;
```

Comment:
Instead of saving this flag `PluginZipStatus`, create a method/procedure that returns the filepath (folder + "/plugin.xml") to be used as argument to `get_plugin_name()`. This new method may be called `find_plugin_xml()` or `find_plugin_metafile()` or anything better XD.

If the file is in in the zip root folder, the filepath would be "./plugin.xml" or simply "plugin.xml".

As you remembered, user may provide an invalid zip file (without plugin.xml). In this case, the returned filepath could be empty to signalize it.

The way you refactored is more organized than before, but it scan the zip twice searching for the same file (as you did before). :) As I propose to you, `get_plugin_name()` will be reduced a lot ;)